### PR TITLE
Enhancement/3406 store error notifications

### PR DESCRIPTION
**Pull request by Hazlitte**
## Summary

Update to replace legacy error notifications with store based infrastructure

Addresses issue #3406

## Checklist

- [X] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [ ] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [X] I have added a QA Brief on the issue linked above.
- [X] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

File: assets/js/components/ActivateModuleCTA.js
Comment (eugene-manuilov) [inline assets/js/components/ActivateModuleCTA.js:55]: @Hazlitte let's move `type`, `format`, and `title` properties to the `setInternalServerError` action and make them as default values to eliminate a need to repeat them in every component.

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Thanks, @Hazlitte. This is a good start. However there are things that we need to address before we can merge this PR.

File: assets/js/googlesitekit/datastore/site/errors.test.js
Comment (aaemnnosttv) [inline assets/js/googlesitekit/datastore/site/errors.test.js:57]: Please use `it` instead of `test` for consistency.

File: assets/js/googlesitekit/datastore/site/errors.test.js
Comment (aaemnnosttv) [inline assets/js/googlesitekit/datastore/site/errors.test.js:56]: Please also group tests by `actions` and `selectors` as is done in all other tests for data store partials.

File: assets/js/googlesitekit/datastore/site/errors.js
Comment (aaemnnosttv) [inline assets/js/googlesitekit/datastore/site/errors.js:102]: This store doesn't have a `getError` selector so this resolver wouldn't do anything.

File: assets/js/googlesitekit/datastore/site/errors.test.js
Comment (aaemnnosttv) [inline assets/js/googlesitekit/datastore/site/errors.test.js:76]: Please also add tests for `getInternalServerError`.

File: assets/js/googlesitekit/datastore/site/errors.test.js
Comment (aaemnnosttv) [inline assets/js/googlesitekit/datastore/site/errors.test.js:44]: These tests don't the API at all so this part should be unnecessary.

**Review by aaemnnosttv — CHANGES_REQUESTED** [review]
Thanks @Hazlitte A few additional points to address. cc: @eugene-manuilov

File: assets/js/components/notifications/InternalServerError.test.js
Comment (aaemnnosttv) [inline assets/js/components/notifications/InternalServerError.test.js:33]:
```suggestion
```
```suggestion
	//const store = registry.stores[ CORE_SITE ].store;
```

**Comment by Hazlitte** [comment]
@aaemnnosttv @eugene-manuilov Thank you for your reviews.  There is an aspect of the changes in this ticket which are unclear to me from reviewing the AC and IB.  I will submit my progress to this branch, but please don't review it as it will fail tests due to me replacing old code which needs a different approach.  I will request more information on the ticket.

**Comment by eugene-manuilov** [comment]
@googlebot I consent.

**Comment by eugene-manuilov** [comment]
@aaemnnosttv could you please take another look at this PR? It should be good now.

File: assets/js/components/legacy-notifications/data-error.js
Comment (aaemnnosttv) [inline assets/js/components/legacy-notifications/data-error.js:76]: While this is an in-place replacement for what was used before, I don't think this is the right place for such a side-effect. In particular, we want to avoid using the global `Data` registry and always use the provided instance via context.

If we wanted to keep this here, we could instead wrap `cta` with a functional component that used `useDispatch` internally. It should probably also use `useEffectOnce` to avoid dispatching this on every render. We could also use a proper selector for the `connectURL` as well. This would fix the global access of the registry and remove the side-effect in the datastore from calling this function.

Alternatively, we could move this logic into `withData`, particularly into `NewComponent::handleReturnedData`. We would need to also wrap `NewComponent` with `withDispatch` to provide `setInternalServerError` as a prop.

File: assets/js/components/data/index.js
Comment (aaemnnosttv) [inline assets/js/components/data/index.js:182]: 👍

File: assets/js/components/legacy-notifications/index.js
Comment (aaemnnosttv) [reply assets/js/components/legacy-notifications/index.js:53]: @eugene-manuilov – just to confirm, the `UnsatisfiedScopesAlert` is still rendered in the `ErrorNotifications` component which is no longer going to be guarded by the feature flag so I think this is okay now, correct?

File: assets/js/components/data/index.js
Comment (aaemnnosttv) [reply assets/js/components/data/index.js:204]: @eugene-manuilov – `AuthError` and `UnsatisfiedScopesAlert` are already present in the new `ErrorNotifications` component which will be the only version of these going forward, however I don't think `setAuthError` is currently called for legacy data requests. Is this handled somewhere?

**Review by aaemnnosttv — CHANGES_REQUESTED** [review]
Thanks @Hazlitte and @eugene-manuilov. This is looking good, just a few things left to address.

File: assets/js/components/legacy-notifications/index.js
Comment (eugene-manuilov) [reply assets/js/components/legacy-notifications/index.js:53]: @aaemnnosttv yes, you are right.

File: assets/js/components/data/index.js
Comment (eugene-manuilov) [reply assets/js/components/data/index.js:204]: Yes, it's handled here: https://github.com/google/site-kit-wp/blob/f538d84956e04a1b9ef2650da6570b6cc0fe97d9/assets/js/googlesitekit/api/index.js#L88-L101

File: assets/js/googlesitekit/datastore/site/errors.js
Comment (eugene-manuilov) [reply assets/js/googlesitekit/datastore/site/errors.js]: Updated.

File: assets/js/googlesitekit/datastore/site/errors.js
Comment (eugene-manuilov) [reply assets/js/googlesitekit/datastore/site/errors.js]: Updated.

File: assets/js/googlesitekit/datastore/site/errors.test.js
Comment (eugene-manuilov) [reply assets/js/googlesitekit/datastore/site/errors.test.js]: Updated.

File: assets/js/components/notifications/InternalServerError.test.js
Comment (eugene-manuilov) [reply assets/js/components/notifications/InternalServerError.test.js]: The `InternalServerError` component seems to be completely unnecessary because it is just a wrapper around the `Notification` component. I have moved the `Notification` component rendering into the `ErrorNotifications` component and removed the `InternalServerError` component and its snapshots completely.

File: assets/js/components/notifications/InternalServerError.test.js
Comment (eugene-manuilov) [reply assets/js/components/notifications/InternalServerError.test.js]: Not applicable anymore.

File: assets/js/components/notifications/InternalServerError.js
Comment (eugene-manuilov) [reply assets/js/components/notifications/InternalServerError.js]: Not applicable anymore.

File: assets/js/components/notifications/ErrorNotifications.js
Comment (eugene-manuilov) [reply assets/js/components/notifications/ErrorNotifications.js]: Updated.

File: assets/js/components/legacy-notifications/data-error.js
Comment (eugene-manuilov) [reply assets/js/components/legacy-notifications/data-error.js:76]: @aaemnnosttv I agree with you, this is not perfect, but it's a temporary solution for legacy stuff that will go away soon. Maybe we can leave it as-is for now? I just don't to spend time on rewriting `withData` because it will be a much bigger change that is out of scope for this ticket.

File: assets/js/components/notifications/ErrorNotifications.js
Comment (aaemnnosttv) [inline assets/js/components/notifications/ErrorNotifications.js:41]: Please restore the `InternalServerError` component as per the ACs. It is a very simple component but let's keep it independent for now as defined.

File: assets/js/components/legacy-notifications/data-error.js
Comment (aaemnnosttv) [reply assets/js/components/legacy-notifications/data-error.js:76]: Sure, I think the `withData` change would be a bit more out of scope, but what about other idea?

I realize it's temporary but we should avoid reaching for `Data.select` and `Data.dispatch` unless there's really no other way.

<details>
<summary>wrapping CTA with error effect component</summary>

```diff
diff --git a/assets/js/components/legacy-notifications/data-error.js b/assets/js/components/legacy-notifications/data-error.js
index 173255f65..9f63dc94a 100644
--- a/assets/js/components/legacy-notifications/data-error.js
+++ b/assets/js/components/legacy-notifications/data-error.js
@@ -20,19 +20,21 @@
  * WordPress dependencies
  */
 import { __, sprintf } from '@wordpress/i18n';
+import { useEffectOnce } from 'react-use';
 
 /**
  * Internal dependencies
  */
 import Data from 'googlesitekit-data';
 import { CORE_SITE } from '../../googlesitekit/datastore/site/constants';
+import { CORE_USER } from '../../googlesitekit/datastore/user/constants';
 import { getModulesData } from '../../util';
 import { isInsufficientPermissionsError } from '../../util/errors';
 import { getInsufficientPermissionsErrorDescription } from '../../util/insufficient-permissions-error-description';
 import ErrorText from '../ErrorText';
 import CTA from './cta';
 import ctaWrapper from './cta-wrapper';
-const { dispatch } = Data;
+const { useSelect, useDispatch } = Data;
 
 /**
  * Creates a CTA component when there's a data error. Different wrapper HTML is needed depending on where the CTA gets output, which is determined by the inGrid, fullWidth, and createGrid parameters.
@@ -63,17 +65,25 @@ function getDataErrorComponent( moduleSlug, errorMessage, inGrid = false, fullWi
 
 	const reconnectURL = errorObj?.data?.reconnectURL;
 	const description = reconnectURL ? <ErrorText message={ message } reconnectURL={ reconnectURL } /> : message;
-	const cta = <CTA title={ title } description={ description } error />;
+	let cta = <CTA title={ title } description={ description } error />;
 
 	// This is to handle token expired error specifically.
 	if ( 'Invalid Credentials' === errorMessage ) {
-		dispatch( CORE_SITE ).setInternalServerError( {
-			id: 'token-notification',
-			title: __( 'Security Token Error', 'google-site-kit' ),
-			description: __( 'We’re unable to retrieve your data because your security token is expired or revoked.', 'google-site-kit' ),
-			learnMoreURL: global._googlesitekitLegacyData.admin.connectURL,
-			learnMoreLabel: __( 'Please reauthenticate your account', 'google-site-kit' ),
-		} );
+		const CTAWithErrorEffect = () => {
+			const connectURL = useSelect( ( select ) => select( CORE_USER ).getConnectURL() );
+			const { setInternalServerError } = useDispatch( CORE_SITE );
+			useEffectOnce( () => {
+				setInternalServerError( {
+					id: 'token-notification',
+					title: __( 'Security Token Error', 'google-site-kit' ),
+					description: __( 'We’re unable to retrieve your data because your security token is expired or revoked.', 'google-site-kit' ),
+					learnMoreURL: connectURL,
+					learnMoreLabel: __( 'Please reauthenticate your account', 'google-site-kit' ),
+				} );
+			} );
+			return cta;
+		};
+		cta = <CTAWithErrorEffect />;
 	}
 
 	return ctaWrapper( cta, inGrid, fullWidth, createGrid );
```

</details>

File: assets/js/components/notifications/ErrorNotifications.js
Comment (eugene-manuilov) [reply assets/js/components/notifications/ErrorNotifications.js]: Restored.

File: assets/js/components/legacy-notifications/data-error.js
Comment (eugene-manuilov) [reply assets/js/components/legacy-notifications/data-error.js]: Ok, I think it can work. Updated, but not sure how to test it :thinking:

File: assets/js/components/data/index.js
Comment (aaemnnosttv) [reply assets/js/components/data/index.js:204]: @eugene-manuilov this only covers requests made with the newer API module. The removed code above was specific to the legacy dataAPI which is still used until removed in #2258.

File: assets/js/components/legacy-notifications/data-error.js
Comment (aaemnnosttv) [reply assets/js/components/legacy-notifications/data-error.js]: @eugene-manuilov I just saw that @felixarntz said this could be completely removed on the issue https://github.com/google/site-kit-wp/issues/3406#issuecomment-857116728 🤷‍♂️ 

I've never seen this error before. I'm guessing it's only possible if you were to manually provide an invalid oauth client ID and secret which is more of an edge case for developers or special environments.

File: assets/js/components/data/index.js
Comment (aaemnnosttv) [reply assets/js/components/data/index.js:204]: Maybe this is okay if we're not concerned about the legacy API any more. @felixarntz ?

File: assets/js/components/legacy-notifications/data-error.js
Comment (aaemnnosttv) [inline assets/js/components/legacy-notifications/data-error.js:76]: Looks like this may be able to be removed entirely as per @felixarntz : https://github.com/google/site-kit-wp/issues/3406#issuecomment-857116728

File: assets/js/components/data/index.js
Comment (aaemnnosttv) [reply assets/js/components/data/index.js:204]: @felixarntz – do we need to ensure these are still handled for legacy data requests?

**Review by aaemnnosttv — COMMENTED** [review]
This looks good to me! A few questions remain for @felixarntz so I'll leave those for him.

File: assets/js/components/data/index.js
Comment (felixarntz) [reply assets/js/components/data/index.js:204]: Let's remove those pieces, we shouldn't need to worry about the legacy data API anymore due to #2258. However, I think it's reasonable to say we shouldn't ship this as a "broken" state. I'm going to assign this issue as well as #2258 to the 1.38.0 release, so that we can tie them together. In other words, let's hold off merging this PR until we branched off 1.37.0 into `main`.

File: assets/js/components/legacy-notifications/data-error.js
Comment (felixarntz) [reply assets/js/components/legacy-notifications/data-error.js:76]: @aaemnnosttv @eugene-manuilov Indeed, this can be removed.

Comment (eugene-manuilov) [reply assets/js/components/legacy-notifications/data-error.js]: Ok, deleted.

**Review by aaemnnosttv — CHANGES_REQUESTED** [review]
LGTM ✅ 

Requesting changes here to make it clear this is waiting for develop to be open for 1.38.0, otherwise this is good to go.

**Comment by aaemnnosttv** [comment]
@eugene-manuilov would you please resolve the conflict?

**Comment by eugene-manuilov** [comment]
@aaemnnosttv fixed.

**Review by aaemnnosttv — APPROVED** [review]
Great, thanks all!

**Comment by aaemnnosttv** [comment]
Looks like there's one last conflict that I'll quickly fix now.

**Comment by aaemnnosttv** [comment]
@googlebot I consent.

**Pull request merged**
